tools/clang/lib/SPIRV/EmitSPIRVAction.cpp

```diff
@@ -49,11 +63,268 @@ spv::ExecutionModel getSpirvShaderStageFromHlslProfile(const char *profile) {
+/// translated into an interface variable; otherwise, we call it as *normal*.
+class DeclResultIdMapper {
+public:
+  DeclResultIdMapper(spirv::ModuleBuilder *builder) : theBuilder(*builder) {}
+
+  /// \brief Defines a function return value in this mapper and returns the
+  /// <result-id> for the final return type.
+  ///
+  /// The final return type is the "residual" type after "stripping" all
+  /// subtypes will attached semantics. For exmaple, stripping "float4 :
+  /// SV_Target" will result in "void", and stripping "struct { float4 :
+  /// SV_Target, float4 }" will result in "struct { float4 }".
+  ///
+  /// Proper SPIR-V instructions will be generated to create the corresponding
+  /// interface variable if stripping happens.
+  uint32_t defineFnReturn(FunctionDecl *funcDecl) {
+    // SemanticDecl for the return value is attached to the FunctionDecl.
+    const auto sk = getInterfaceVarSemanticAndKind(funcDecl);
+    if (sk.second != InterfaceVariableKind::None) {
+      // Found return value with semantic attached. This means we need to map
+      // the return value to a single interface variable.
+      const uint32_t retTypeId =
+          translateType(funcDecl->getReturnType(), theBuilder);
+      // TODO: Change to the correct interface variable kind here.
+      const uint32_t varId = theBuilder.addStageIOVariable(
+          retTypeId, spv::StorageClass::Output, llvm::None);
+
+      stageOutputs.push_back(std::make_pair(varId, sk.first));
+      remappedDecls[funcDecl] = varId;
+      interfaceVars.insert(varId);
+
+      return theBuilder.getVoidType();
+    } else {
+      // TODO: We need to handle struct return types here.
+      return translateType(funcDecl->getReturnType(), theBuilder);
+    }
+  }
+
+  /// \brief Defines a function parameter in this mapper and returns the
+  /// <result-id> for the final parameter type. Returns 0 if the final type
+  /// is void.
+  ///
+  /// The final parameter type is the "residual" type after "stripping" all
+  /// subtypes will attached semantics. For exmaple, stripping "float4 :
+  /// SV_Target" will result in "void", and stripping "struct { float4 :
+  /// SV_Target, float4 }" will result in "struct { float4 }".
+  ///
+  /// Proper SPIR-V instructions will be generated to create the corresponding
+  /// interface variable if stripping happens.
+  uint32_t defineFnParam(ParmVarDecl *paramDecl) {
+    const auto sk = getInterfaceVarSemanticAndKind(paramDecl);
+    if (sk.second != InterfaceVariableKind::None) {
+      // Found parameter with semantic attached. This means we need to map the
+      // parameter to a single interface variable.
+      const uint32_t paramTypeId =
+          translateType(paramDecl->getType(), theBuilder);
+      // TODO: Change to the correct interface variable kind here.
+      const uint32_t varId = theBuilder.addStageIOVariable(
+          paramTypeId, spv::StorageClass::Input, llvm::None);
```

Comment:
Yep, it truly depends. Not all parameters are inputs to the shader. I've TODO in the above.